java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -365,46 +403,32 @@ public SymbolicValue getValue(Symbol symbol) {
     return values.get(symbol);
   }
 
-  public Map<SymbolicValue, ObjectConstraint<ObjectConstraint.Status>> getValuesWithConstraints(final Object state) {
-    final Map<SymbolicValue, ObjectConstraint<ObjectConstraint.Status>> result = new HashMap<>();
-    constraints.forEach((symbolicValue, valueConstraint) -> {
-      if (valueConstraint instanceof ObjectConstraint) {
-        ObjectConstraint<ObjectConstraint.Status> constraint = (ObjectConstraint<ObjectConstraint.Status>) valueConstraint;
-        if (constraint.hasStatus(state)) {
-          result.put(symbolicValue, constraint);
-        }
+  public List<SymbolicValue> getValuesWithConstraints(final Constraint constraint) {
+    final List<SymbolicValue> result = new ArrayList<>();
+    constraints.forEach((symbolicValue, constraintByDomain) -> {
+      Constraint find = constraintByDomain.get(constraint.getClass());
+      if(find == constraint) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)